lib/roby/interface/client.rb

```diff
@@ -131,6 +142,15 @@ def wait(timeout: nil)
                 io.read_wait(timeout: timeout)
             end
 
+            # @api private
+            #
+            # Remove and call the block of a pending async call
+            def process_pending_async_call(error, result)
+                current_call = pending_async_calls.shift
+                call_block = current_call[:block]
+                call_block.call(error, result) if call_block
```

Comment:
I have a funny feeling w.r.t. error handling and the ability to not provide a block.

In this case, I would go for raising 'as-if' the call was synchronous, but modifying the error message to mention the original method call. What do you think ?

Reply:
You mean raising in case of a :bad_call AND if no block was given? The rationale to allow no block and in this case ignore a :bad_call was: if there is no block then the user really doesn't care about the return (maybe the use cases are limited here but heartbeats come to mind, for example). Besides, raising/throwing exceptions in async calls with callbacks is not a very common pattern (at least that I know of) and in this case would force the user to modify his/her event loop (Qt's timer timeout() slot in our case) to catch and deal with the error which maybe completely outside of the context/scope of what he/she is doing when using this #async_call. So, personally, I would say no raising if no block is given (unless #async_call fails before anything is written to the remote side like when an action doesn't exist).

That all being said, I am not very familiar with roby, its intricacies and uses in the wild so I will gladly follow your lead here... :)

Reply:
I think that's all good points. I don't think there's a one-size-fit-all policy, that's why I wanted your opinion.

The thing I'm generally against is to silently ignore unhandled exceptions. If a callback decides to ignore it, then fine it's the developer's choice. However, allowing the caller to not provide any callback *and* having no notification of an exception sounds like a recipe for disaster.

Roby's promise system for instance will queue an unhandled exception as a general rror if it's not handled. However, this is not the primary system, but the external interface, which might relax the rules a bit. 

To give some context, :bad_call is really an error that happened during the call processing (i.e. not a communication error), which means that the caller should simply not have called this method at this point (or the  arguments are wrong, or ... you get the idea). Moreover, throwing an exception at this point has the effect of terminating the connection and triggering a reconnection.

Generally speaking, I think I would be fine with at least logging the exception at first, and see if that's enough to not make the developer's life hell. I can tell you that the JS default policy of simply ignoring promise errors is horrible (I've been doing some of that the last weeks)..

Reply:
Well, it sounds then that ignoring the exception goes against roby's philosophy and that is usually a bad idea from a design point of view... In which case it is probably better to not allow a nil block. I mean, there's virtually no boilerplate involved in passing an empty block if the developer explicitly wants to ignore the exception (`async_call(...) { }`). I think that's a sane middle ground (what I really dislike is the idea of raising in a context where the developer might not even be able to catch it). Agree?

Reply:
Agreed

Reply:
:+1: